src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs

```diff
@@ -162,6 +190,69 @@ protected override void ExecuteCore()
             PlatformManifests = platformManifests.ToArray();
             PackageConflictOverrides = packageConflictOverrides.ToArray();
             PackageConflictPreferredPackages = string.Join(";", preferredPackages);
+
+            if (s_allowCacheLookup)
+            {
+                ResolvedAssetsCacheEntry newCacheEntry = new()
+                {
+                    ReferencesToAdd = ReferencesToAdd,
+                    AnalyzersToAdd = AnalyzersToAdd,
+                    PlatformManifests = PlatformManifests,
+                    PackageConflictPreferredPackages = PackageConflictPreferredPackages,
+                    PackageConflictOverrides = PackageConflictOverrides,
+                    UsedRuntimeFrameworks = UsedRuntimeFrameworks,
+                };
+    
+                BuildEngine4?.RegisterTaskObject(cacheKey, newCacheEntry, RegisteredTaskObjectLifetime.AppDomain, allowEarlyCollection: true);
+            }
+
+        }
+
+        /// <summary>
+        /// Construct a key for the cache lookup.
+        /// </summary>
+        /// <remarks>
+        /// IMPORTANT: any input changes that can affect the output should be included in this key.
+        /// </remarks>
+        /// <returns>A string suitable for cache lookups based on the inputs of the task.</returns>
+        internal string OverallCacheKey()
+        {
+            StringBuilder cacheKeyBuilder = new(nameof(ResolveTargetingPackAssets) + nameof(OverallCacheKey));
+            cacheKeyBuilder.AppendLine();
+
+            cacheKeyBuilder.AppendLine(nameof(FrameworkReferences));
+            foreach (var frameworkReference in FrameworkReferences)
+            {
+                cacheKeyBuilder.AppendLine(frameworkReference.ItemSpec);
+            }
+            cacheKeyBuilder.AppendLine();
+            cacheKeyBuilder.AppendLine(nameof(ResolvedTargetingPacks));
+            foreach (var resolvedTargetingPack in ResolvedTargetingPacks)
+            {
+                cacheKeyBuilder.AppendLine(resolvedTargetingPack.ItemSpec);
+                cacheKeyBuilder.AppendLine(resolvedTargetingPack.GetMetadata(MetadataKeys.Path));
+                cacheKeyBuilder.AppendLine(resolvedTargetingPack.GetMetadata("TargetingPackFormat"));
+                cacheKeyBuilder.AppendLine(resolvedTargetingPack.GetMetadata("TargetFramework"));
+                cacheKeyBuilder.AppendLine(resolvedTargetingPack.GetMetadata(MetadataKeys.NuGetPackageId));
+                cacheKeyBuilder.AppendLine(resolvedTargetingPack.GetMetadata(MetadataKeys.NuGetPackageVersion));
+                cacheKeyBuilder.AppendLine(resolvedTargetingPack.GetMetadata(MetadataKeys.PackageConflictPreferredPackages));
```

Comment:
Ideally the tests would verify that the key depends on any metadata items that are read.  The way to do that would probably be to use a strongly-typed wrapper for all metadata access.